[namprice227] iP

namprice227 opened the pull request:
# Charlie
New type of **Chatbot** that:
- Help you remember task :grinning:
- Got features deadline, event, todo
- Friendly chatbot

What you need to do?
1.  Download the latest release from here [Charlie](https://github.com/namprice227/ip/releases/tag/A-Jar)
2. Run it with  
```java -jar charlie.jar```
4. Very beginner-friendly, type anything for help

It is similar to you ticking your own tickbox:
- [ ] deadline cs2103t home work /by 2025/02/06
- [x] todo housework

If you want to programmer remember to include this:
``` java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

namprice227 committed:
Create chatbot Charlie

namprice227 committed:
Add task is available

namprice227 committed:
Save changes before switching branches

namprice227 committed:
Update with mark feature

namprice227 committed:
Finish subtask like todo, deadline, event

namprice227 committed:
Add testing cases, but the script may have problem?

namprice227 committed:
Fix bugs regarding io

namprice227 committed:
Add some handler and instructions to help it more usable

namprice227 committed:
Add delete features

namprice227 committed:
Contain more handlers for Deadline and Event

namprice227 committed:
Add extra IO test

namprice227 committed:
Merge branch 'master' of https://github.com/namprice227/ip

namprice227 committed:
Read from file for number of tasks, add some write features

namprice227 committed:
Add write to file features but not delete not mark or unmark

namprice227 committed:
Fix bugs  write to file features

namprice227 committed:
Change the way to add tasks to file, now after each action, all the tasks are written again to the file

namprice227 committed:
Fix bugs to add to files

namprice227 committed:
Merge branch 'branch-Level-7'

namprice227 committed:
Add date-time function

namprice227 committed:
Merge branch 'branch-Level-8'

namprice227 committed:
Add Parser, Storage, TaskList, Ui class and make the app much more OOP

namprice227 committed:
Finish OOP code

namprice227 committed:
Merge branch 'A-MoreOOP'

namprice227 committed:
Create package

namprice227 committed:
Merge branch 'add-gradle-support'

namprice227 committed:
Add gradle

namprice227 committed:
Build some Junits for class TaskList and Deadline

namprice227 committed:
Create feature for jar file

namprice227 committed:
Add javadoc comment

namprice227 committed:
Try my best to make it coding standard

namprice227 committed:
Add find features

namprice227 committed:
Merge branch 'branch-A-JavaDoc'

namprice227 committed:
Fix conflict on Parser.java

namprice227 committed:
Fix merge conflict

huangtian666 commented on src/main/java/charlie/Charlie.java:
```diff
@@ -0,0 +1,30 @@
+ * It initializes storage, loads tasks, and processes user commands.
+ */
+
+public class Charlie {
```
I like how you have coded in a more OOP way like separating into different components such as UI, Task and so on!

namprice227 replied:
Thank you =)

huangtian666 commented on src/main/java/charlie/Charlie.java:
```diff
@@ -0,0 +1,30 @@
+        this.taskList = new TaskList(storage);
+    }
+
+    public void run() {
+        Ui.showWelcome();
+        while (true) {
+            try {
+                String fullCommand = Ui.readCommand();
+                Parser parser = new Parser(fullCommand);
+                parser.action(taskList);
+            } catch (Exception e) {
+                Ui.showError(e.getMessage());
+            }
+        }
+    }
```
It is nice to see that your implementation is so clean and simple!

huangtian666 commented on src/main/java/charlie/Ui.java:
```diff
@@ -0,0 +1,51 @@
+    /**
+     * Prints a help message that shows the available commands the user can use.
+     */
+    public static void printHelp() {
+        System.out.println("Here are the commands you can use:");
+        System.out.println("1. todo <task description>");
+        System.out.println("2. deadline <task description> /by <time>");
+        System.out.println("3. event <event description> /from <start time> /to <end time>");
+        System.out.println("4. delete <task number>");
+        System.out.println("5. list");
+        System.out.println("6. bye");
+    }
```
Your help method is really helps to improve the user experience!

huangtian666 commented on src/main/java/charlie/Parser.java:
```diff
@@ -0,0 +1,83 @@
+                exit(0);
+            }
+            default -> Ui.printHelp();
+        }
```
Your choice of switch statements and lambda expression have effectively simplified the code and increased readability for readers!

huangtian666 commented on src/main/java/charlie/Charlie.java:
```diff
@@ -0,0 +1,30 @@
+package charlie;
```
One suggestion is that you could split the files into more packages so all relevant components are in the same folder!

TheEnd-alr-taken replied:
To add on to what @huangtian666 mentioned, I think it will look neater if you could create a charlie.task package and group up all the relevant components together since you have already taken the time to extract task specific classes.

namprice227 replied:
Thank you for your suggestion, I will consider this

huangtian666 commented on src/main/java/charlie/Todo.java:
```diff
@@ -0,0 +1,25 @@
+    public Todo(String activity, Boolean marked) {
+        super(activity, marked);
+    }
+
```
One minor suggestion here is to include @Override when you override a method in the superclass 😊

huangtian666 reviewed (commented):
Overall, nice implementation so far!

namprice227 committed:
Make GUI features

namprice227 committed:
Change minimal

namprice227 committed:
Fix some bits

TheEnd-alr-taken commented on src/main/java/charlie/CharlieHandler.java:
```diff
@@ -0,0 +1,94 @@
+     * @param words The Deadline task description.
+     * @return {@code true} if the description is valid, {@code false} otherwise.
+     */
+
```
I feel that the blank line between the documentation block and the method could be omitted to standardise with other areas of the code.

TheEnd-alr-taken commented on src/main/java/charlie/Parser.java:
```diff
@@ -0,0 +1,154 @@
+            }
+            case "mark" -> {
+                if (CharlieHandler.isValidNumber(task)) {
+                    taskList.markTask(parseInt(task));
+                }
+            }
+            case "unmark" -> {
+                if (CharlieHandler.isValidNumber(task)) {
+                    taskList.unmarkTask(parseInt(task));
+                }
+            }
+            case "todo" -> {
+                if (CharlieHandler.isValidTodo(task)) {
+                    taskList.addTask(new Todo(task));
+                }
+            }
+            case "deadline" -> {
+                if (CharlieHandler.isValidDeadline(task)) {
+                    taskList.addTask(new Deadline(task));
+                }
+            }
+            case "event" -> {
+                if (CharlieHandler.isValidEvent(task)) {
+                    taskList.addTask(new Event(task));
+                }
+            }
+            case "delete" -> {
+                if (CharlieHandler.isValidNumber(task)) {
+                    taskList.deleteTask(parseInt(task));
+                }
+            }
+            case "find" -> {
+                taskList.findTask(task.trim());
+            }
+            case "bye" -> {
+                Ui.showGoodbye();
+                exit(0);
+            }
+            default -> Ui.printHelp();
+        }
```
As mentioned by @huangtian666 , I also like the choice of using lambda expression here. Any reason why you choose to separate this action(TaskList taskList) method from the getResponse(TaskList taskList) method? Both seems to have the same commands being run except for the returning of the responses.

namprice227 replied:
Oh, I was implementing the GUI, so I need to return the string instead of printing it out. I think I will delete the action method in the future as they said can delete the cli

TheEnd-alr-taken reviewed (commented):
Overall, looks good to me. Congrats on making this much progress.

namprice227 committed:
Merge branch 'branch-checkstyle'

namprice227 committed:
mprove Parser input validation and error handling

Add assertions in the Parser constructor and methods to ensure that
critical
inputs are valid before proceeding. In particular, the changes include:

- Asserting that the input string provided to the Parser constructor is
  not
    null. This prevents null pointer exceptions when splitting the
    string.
    - Handling cases where the command may lack additional parameters by
      defaulting
        the task field to an empty string.
	- Adding assertions in the action() and getResponse() methods to
	  verify that the
	    TaskList parameter is not null, ensuring that downstream
	    methods operate on
	      valid objects.

	      These improvements enhance the robustness of the command
	      parsing logic and
	      help catch programming errors early during development.

namprice227 committed:
Improve Parser input validation and error handling

Add assertions in the Parser constructor and methods to ensure that
critical
inputs are valid before proceeding. In particular, the changes include:

- Asserting that the input string provided to the Parser constructor is not null. This prevents null pointer exceptions when splitting the string.
- Handling cases where the command may lack additional parameters by defaulting the task field to an empty string.
- Adding assertions in the action() and getResponse() methods to verify that the TaskList parameter is not null, ensuring that downstream methods operate on valid objects.

namprice227 committed:
Delete some unnecessary methods, factor some methods outside

namprice227 committed:
Merge pull request #1 from namprice227/branch-A-Assertions

Request to add some assertions

namprice227 committed:
Merge branch 'branch-A-CodeQuality' of https://github.com/namprice227/ip

namprice227 committed:
Merge branch 'master' into branch-A-CodeQuality

namprice227 committed:
Merge pull request #2 from namprice227/branch-A-CodeQuality

Request to improve the code quality

namprice227 committed:
Add accidentally delete file

namprice227 committed:
Finish the features

namprice227 committed:
Finish the find feature

namprice227 committed:
Merge branch 'branch-C-BetterSearch'

namprice227 committed:
Add github actions

namprice227 committed:
Disable checkstyle to run the code

namprice227 committed:
Delete I/O redirection of gradle.yml file

namprice227 committed:
Use AI to fix some codes

namprice227 committed:
Change all Duke to Charlie

namprice227 committed:
Add initial message

namprice227 committed:
Add readMe file and the pic

namprice227 committed:
Add image to readme

namprice227 committed:
Add pic

namprice227 committed:
Fix some minor bugs

namprice227 committed:
Add jar file

namprice227 committed:
Add jar header

namprice227 committed:
Merge branch 'master' of https://github.com/namprice227/ip

namprice227 committed:
Add files

namprice227 committed:
Try to fix Linux fault

namprice227 committed:
Fix something

namprice227 committed:
Fix absolute path problem
